text/0000-inline-asm.md

```diff
@@ -0,0 +1,892 @@
+among others can be addressed by their name.
+
+```rust
+unsafe {
+    asm!("out 0x64, {}", in("eax") cmd);
+}
+```
+
+In this example we call the `out` instruction to output the content of the `cmd` variable
+to port `0x64`. Since the `out` instruction only accepts `eax` (and its sub registers) as operand
+we had to use the `eax` constraint specifier.
+
+It is somewhat common that instructions have operands that are not explicitly listed in the
+assembly (template). Hence, unlike in regular formatting macros, we support excess arguments:
+
+```rust
+fn mul(a: u32, b: u32) -> u64 {
+    let lo: u32;
+    let hi: u32;
+
+    unsafe {
+        asm!(
+            // The x86 mul instruction takes eax as an implicit input and writes
+            // the 64-bit result of the multiplication to eax:edx.
+            "mul {}",
+            in(reg) a, in("eax") b,
+            lateout("eax") lo, lateout("edx") hi
+        );
+    }
+
+    hi as u64 << 32 + lo as u64
+}
+```
+
+This uses the `mul` instruction to multiply two 32-bit inputs with a 64-bit result.
+The only explicit operand is a register, that we fill from the variable `a`.
+The second operand is implicit, and must be the `eax` register, which we fill from the variable `b`.
+The lower 32 bits of the result are stored in `eax` from which we fill the variable `lo`.
+The higher 32 bits are stored in `edx` from which we fill the variable `hi`.
+
+Note that `lateout` must be used for `eax` here since we are specifying the same register as both an input and an output.
+
+## Clobbered registers
+
+In many cases inline assembly will modify state that is not needed as an output.
+Usually this is either because we have to use a scratch register in the assembly,
+or instructions modify state that we don't need to further examine.
+This state is generally referred to as being "clobbered".
+We need to tell the compiler about this since it may need to save and restore this state
+around the inline assembly block.
+
+```rust
+let ebx: u32;
+let ecx: u32;
+
+unsafe {
+    asm!(
+        "cpuid",
+        in("eax") 4, in("ecx") 0,
+        lateout("ebx") ebx, lateout("ecx") ecx,
+        lateout("eax") _, lateout("edx") _
+    );
+}
+
+println!(
+    "L1 Cache: {}",
+    ((ebx >> 22) + 1) * (((ebx >> 12) & 0x3ff) + 1) * ((ebx & 0xfff) + 1) * (ecx + 1)
+);
+```
+
+In the example above we use the `cpuid` instruction to get the L1 cache size.
+This instruction writes to `eax`, `ebx`, `ecx`, and `edx`, but for the cache size we only care about the contents of `ebx` and `ecx`.
+
+However we still need to tell the compiler that `eax` and `edx` have been modified so that it can save any values that were in these registers before the asm. This is done by declaring these as outputs but with `_` instead of a variable name, which indicates that the output value is to be discarded.
```

Comment:
I'd prefer if things like clobbering could be motivated and specified in terms of the constraints that are put *on the assembly*, not on the compiler. I gather people writing inline assembly are used to thinking very much in terms of compiler details such as register allocation, but I'd prefer if any such remarks would be non-normative notes.

For example, I'd say that the motivation for explicit declaration of clobbered registers is that assembly may not access (read nor write) any register not declared in its inputs and outputs. So, the `cpuid` example without the `lateout("eax") _, lateout("edx") _` is UB due to violating those constraints. That motivates clobbering without ever mentioning what the compiler ends up doing with this information. The fact that the compiler has to save and restore registers is a compiler implementation detail that follows from this description, but it should not be normative.

Phrasing things in terms of constraints on the compiler is akin to specifying a language feature by the set of allowed optimizations: it leaves us with no way of understanding the feature as a language construct on its own right, independent of compilation or any such thing. This is the same concern as trying to specify `black_box` as "inhibiting optimizations" -- neither optimizations nor things like register save/restore should occur in a language spec, IMO.

From what I can tell, basically everything in here *can* be described in the abstract way I propose, which makes me very happy! As a formal methods person, inline assembly is something like my Kryptonite, but this RFC is very well-written and I felt I could mostly follow it. :D   It's just a matter of wording, I think.  The one point I am less sure about is the `lateout` thing... is it enough to say "if any input register gets read after this output was written to, that's UB"?

Reply:
This part is in the guide-level explanation, which tries to be a gentle (ish) introduction to inline assembly rather than a proper specification. In this particular case, it helps to explain to users why indicating proper clobbers is important. The reference-level explanation is more rigorous in specifying exactly what is/isn't UB.

Now regarding the `lateout` thing... I wouldn't say that behavior is undefined if you misuse it, just that you may end up (depending on the register allocator's choices) accidentally overwriting one of your inputs. The problem is entirely confined to the asm code and does not cause any UB and the Rust abstract machine or LLVM level. Hence the looser wording ("should" instead of "must").

Reply:
> This part is in the guide-level explanation, which tries to be a gentle (ish) introduction to inline assembly rather than a proper specification. In this particular case, it helps to explain to users why indicating proper clobbers is important.

I agree that it helps to motivate why something is UB. But even for the gentle introduction, I think it would help to frame this as "that code is UB (and here's why), and here's how we add these clobber annotations to fix that". It puts the reader into the right frame of mind. Otherwise, we risk educating people to primarily think about the compiler's choices when writing such code.

The way things stand right now, the guide does not ever say anything about UB. That is quite unusual for Rust where we are usually very explicit, even in guide-level material, about what the rules are to avoid UB.

> The reference-level explanation is more rigorous in specifying exactly what is/isn't UB.

The term "UB"/"Undefined Behavior" does not even occur in the RFC, though. So, what you are saying is not entirely accurate.
I also looked for the clause that says that reading or writing any register not explicitly specified is UB, and could not find that either.

> Now regarding the lateout thing... I wouldn't say that behavior is undefined if you misuse it, just that you may end up (depending on the register allocator's choices) accidentally overwriting one of your inputs.

If misusing lateout is not UB, we have to specify what happens in that case. So, what does that spec look like? "Non-deterministically, writing to a lateout register might overwrite an input register"?

> Hence the looser wording ("should" instead of "must").

It is far from clear that this is a deliberate choice. I am not sure what "should" means in a spec. Either behavior is UB or it is specified. The spec isn't really the right place for recommendations or advice.

Reply:
For lateout, the specification is:

> the register allocator can reuse a register allocated to an `in`.

Everything else follows as a consequence of that: if you write to the output, you may overwrite one of your inputs.

Reply:
> Now regarding the lateout thing... I wouldn't say that behavior is undefined if you misuse it, just that you may end up (depending on the register allocator's choices) accidentally overwriting one of your inputs. The problem is entirely confined to the asm code and does not cause any UB and the Rust abstract machine or LLVM level. Hence the looser wording ("should" instead of "must").

On some level it depends on whether we define this as undefined behavior. I'd rather do so, because breaking this invariant can trivially cause undefined behavior in otherwise well-defined (assembly) code.
Consider for example:
```rust
let xs = [1u32, 2];
let mut sum;
asm!("
    mov {sum41}, 41
    sum_loop:
        add {sum41}, [{ptr} + ecx*4 - 1]
    loop sum_loop
    ",
    sum41 = lateout(reg) sum,
    ptr = in(reg) xs.as_ptr(),
    in("ecx") xs.len(),
);
assert_eq!(sum, 44);
```
Whenever the `lateout` shadows either `in` this causes out-of-bounds accesses.

Reply:
Is out-of-bounds access within an asm block even considered undefined behavior in "the compiler may optimize assuming this code is never executed at all" meaning? I would expect a much weaker assumption that the compiler can just assume that the out-of-bounds accesses don't touch anything rustc knows/cares about.

> For example, I'd say that the motivation for explicit declaration of clobbered registers is that assembly may not access (read nor write) any register not declared in its inputs and outputs.

This is way too strong. In x86, the NOP instruction literally means "swap the contents of the eax register with the contents of eax", but it would be silly to require any assembly block with a nop in it to mark eax as a clobber. In general reading registers is always harmless and writing them only matters if their value isn't restored later, so it seems like it should be enough to say that it is undefined behavior to modify (and not subsequently restore) a register unless it is listed in an output/clobber.

>  The one point I am less sure about is the lateout thing... is it enough to say "if any input register gets read after this output was written to, that's UB"?

Why do you need to say anything? I'd expect all the register constraints to be in terms of which assignments can alias with which other assignments, and lateout would have the property of needing a register which is distinct from any other output register or clobber, but possibly aliasing an input register. Reasoning about the assembly block would then come down to considering how it behaves under each possible aliasing combination.

Reply:
> For example, I'd say that the motivation for explicit declaration of clobbered registers is that assembly may not access (read nor write) any register not declared in its inputs and outputs.

This is wrong though. For instance, the following piece of inline asm does not clobber eax:

```rust
asm!("push eax; mov eax, 4; pop eax");
```

Reply:
> Is out-of-bounds access within an asm block even considered undefined behavior 

No, this is not undefined behavior, e.g., 

```rust
let ptr: *mut f32 = GlobalAlloc::alloc(Layout::new(8, 16)?) as *mut _;
assert!(ptr as usize % 16 == 0); // aligned to a 16 byte boundary
let mut arr = [MaybeUninit<f32>; 4];
asm!("
        simd_16byte_load xmm0, {0}
        simd_16byte_write {1}, xmm0
    ", 
    in(ptr), out(&arr));
```

allocates 8 bytes on the heap at a 16 byte aligned boundary, and then uses a SIMD load to read 16 byte from the 8-byte wide allocation, reading the last 8-bytes out-of-bounds within the assembly block, and that's ok.

What's UB is an `asm!` block that has a side-effect but does not state it, e.g., an `asm!` block that modifies a register, but does not mark it as being an output or a clobber, since the compiler will generate code under the assumption that this register is not modified, yet that assumption would be incorrect.

Reply:
> This is way too strong. In x86, the NOP instruction literally means "swap the contents of the eax register with the contents of eax", but it would be silly to require any assembly block with a nop in it to mark eax as a clobber. In general reading registers is always harmless and writing them only matters if their value isn't restored later, so it seems like it should be enough to say that it is undefined behavior to modify (and not subsequently restore) a register unless it is listed in an output/clobber.

Thanks for adding that, I was not aware of this.

@gnzlbg's definition makes sense to me.

> This subtly different definition means that an assembly block can actually read from almost any register, including those that are not specified as inputs (e.g. like flags registers), because doing so is not a side-effect. I say almost any, because reading from some registers can be a side-effect (i.e. can modify the value of that or other registers).

It probably also goes without saying that there are no guarantees whatsoever for the value that is being read from such registers.

Reply:
> ("Read" is a kind of effect though, in the sense that entirely pure code ["mathematical functions"] must not even read from mutable things. It's just a harmless effect in many cases. I am not sure how much bearing this fact has on the discussion at hand, but I get a bit worried every time someone claims that reads don't have side-effects.)

I think this is a software (compiler) vs hardware thing. If you read in source code, the compiler has to make sure there's a value there for you to read, which affects the code it generates, meaning there were side-effects to that read - and if the thing you were reading is mutable, this might visibly change what's written there (conversely, if you're a pure function then it might not make sure that you have something to read and then UB). At the hardware level, there's always something to read at that register, even if you can't guarantee what it is, so no extra effort needs to be made. A pure Rust function could have inline asm with the pure marker, and reading from a random register should still be no problem (although, of course, odd). Maybe otherwise put, in the "memory-model" of the cpu-registers, read really isn't an effect (except for the special registers where it is). For actual physical memory, that's probably more delicate.